contracts/contracts/FeeSub.sol

```diff
@@ -0,0 +1,122 @@
+
+import "./RequestManager.sol";
+
+contract FeeSub is Ownable {
+    using SafeERC20 for IERC20;
+
+    RequestManager public requestManager;
+    mapping(address token => uint256) public minimumAmounts;
+    mapping(bytes32 requestId => address sender) public senders;
+
+    constructor(address _requestManager) {
+        requestManager = RequestManager(_requestManager);
+    }
+
+    function createRequest(
+        uint256 targetChainId,
+        address sourceTokenAddress,
+        address targetTokenAddress,
+        address targetAddress,
+        uint256 amount,
+        uint256 validityPeriod
+    ) external returns (bytes32) {
+        require(
+            minimumAmounts[sourceTokenAddress] > 0,
+            "Token not to be subsidized"
+        );
+        require(
+            amount >= minimumAmounts[sourceTokenAddress],
+            "Transfer amount too small to be subsidized"
+        );
+
+        require(
+            IERC20(sourceTokenAddress).allowance(msg.sender, address(this)) >=
+                amount,
+            "Insufficient allowance"
+        );
+
+        IERC20(sourceTokenAddress).safeTransferFrom(
+            msg.sender,
+            address(this),
+            amount
+        );
+
+        bytes32 requestId = requestManager.createRequest(
+            targetChainId,
+            sourceTokenAddress,
+            targetTokenAddress,
+            targetAddress,
+            amount,
+            validityPeriod
+        );
+        senders[requestId] = msg.sender;
+
+        return requestId;
+    }
+
+    function transferableAmount(
+        uint256 targetChainId,
+        address tokenAddress,
+        uint256 amount
+    ) public view returns (uint256) {
+        if (amount >= minimumAmounts[tokenAddress]) {
+            return amount;
+        }
+
+        return
+            requestManager.transferableAmount(
+                targetChainId,
+                tokenAddress,
+                amount
+            );
+    }
+
+    function withdrawExpiredRequest(bytes32 requestId) external {
+        (
+            address sender,
+            address sourceTokenAddress,
+            ,
+            uint256 amount,
+            ,
+            ,
+            ,
+            ,
+            ,
+            ,
+
+        ) = requestManager.requests(requestId);
+        require(
+            sender == address(this),
+            "Request was not created by this contract"
+        );
+        require(senders[requestId] != address(0), "Already Withdrawn");
+
+        requestManager.withdrawExpiredRequest(requestId);
+
+        address recipient = senders[requestId];
+        senders[requestId] = address(0);
+
+        IERC20 token = IERC20(sourceTokenAddress);
+        token.safeTransfer(recipient, amount);
+    }
+
+    function setMinimumAmount(
+        address tokenAddress,
+        uint256 amount
+    ) external onlyOwner {
+        if (amount > 0) {
```

Comment:
We can save some gas here and ask if minimumAmounts[tokenAddress] is already > 0, then we know that we already approved and don't need to approve again.